doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,222 @@
+- A FIPS 140-3 compliant module requires a built-in service indicator capable of indicating the use of approved security services
+- If a module only supports approved services in an approved manner an implicit indicator can be used (e.g. successful completion of a service is an indicator).
+- An approved algorithm is not considered to be an approved implementation if it does not have a CAVP certificate or does not include its required self-tests. (i.e. My interpretation of this is that if the CAVP certificate lists an algorithm with only a subset of key sizes, digests, and/or ciphers compared to the implementation, the differences ARE NOT APPROVED. In many places we have no restrictions on the digest or cipher selected).
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators. AWS/google have added a table in their security policy called ‘Non-Approved Algorithms not allowed in the approved mode of operation’. An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+Since any new FIPS restrictions added could possibly break existing applications
+the following additional OpenSSL requirements are also needed:
+
+- The FIPS restrictions should be able to be disabled using Configuration file options (This results in unapproved mode and requires an indicator).
+- A mechanism for logging the details of any unapproved mode operations that have been triggered (e.g. DSA Signing)
+- The FIPS restrictions should be able to be enabled/disabled per algorithm context.
+- If the per algorithm context value is not set, then the  Configuration file option is used.
+
+Solution
+--------
+
+In OpenSSL most of the existing code in the FIPS provider is using
+implicit indicators i.e. An error occurs if existing FIPS rules are violated.
+
+The following rules will apply to any code that currently is not FIPS approved,
+but needs to be.
+
+- The fipsinstall application will have a configurable item added for each algorithm that requires a change.
+These options will be passed to the FIPS provider in a manner similar to existing code.
+
+- A user defined callback similar to OSSL_SELF_TEST will be added. This callback
+will be triggered whenever an approved mode test fails.
+It may be set up by the user using
+
+```c
+OSSL_INDICATOR_set_callback(OSSL_LIB_CTX *libctx, OSSL_CALLBACK *cb, void *cbarg)
+```
+
+- The callback API will be
+
+```c
+int OSSL_INDICATOR_callback(OSSL_LIB_CTX *libctx, const char *algtype,
+                            const char *algdesc)
+```
```

Comment:
I was thinking that this is just a very generic system for logging strings, not for passing complex context specific types back..
For an example see dsa_sig.c dsa_init_check_approved()
https://github.com/openssl/openssl/pull/23890/files#diff-0c2b4e0a207614d420d8f376c476115e33d5ad7733790a40b1c238a2a79e40c1

In this case the callback may be triggered twice..
Once because dsa signing is not allowed & a second time if the security check fails
Once with "DSA", "Sign Init"
and the second time with "DSA", "Key Size".

Reply:
The cbarg is set via OSSL_INDICATOR_set_callback

Reply:
OSSL_INDICATOR_callback() is actually internal.

Reply:
I have updated the doc a bit

Reply:
thank you for a link to draft-pr looking at the code all is clear now (at least around callback).